### PR TITLE
feat: auto label issues-prs by type & scope

### PR DESCRIPTION
**Pull request by casswedson**
#### Summary
None

#### Purpose of change
Rationale: to add more options to our auto labeling, this lets people
categorize their pr/issue based on the outlines defined on this pr

Also this is overall "pretty basic", it covers most of our regular labels
but it could cover much, much more like every thing about the game I am
not gonna cover all of that now, maybe later

Uses conventional commits' general guidelines to auto label our prs and
issues (see references and testing for examples)

TLDR;
Keep in mind all of this is optional, case sensitive!
conventional commits defines
- mandatory type
- optional scope must be surrounded by `()`
- mandatory colon followed by a space; to help readability
- mandatory short message, communicating what this commit does

`type(scope): short message`

Some examples of CC's guidelines + some things defined here

`fix(ui): remove extra spaces in eat menu`
gets bugfix and user interface labels
`chore: fix typos`
gets chore label
`build(perf): skip duplicates`
gets build and performance labels
`perf: don't lint releases`
gets performance label

some types are also scopes, like ui, performance, etc
some things are limited to scopes like maps/mapgen, tools, i18n etc

also easy to expand!

NOTICE: Chore and CI labels need to be crated beforehand to be used
unsupported till someone makes them, I suggest:

Chore: Updating, grunt tasks etc; no important code changes
CI: Continuous Integration, related to automation for Github, builds and releases

It's best to merge #55828 then #55830 and lastly this, I moved files around,
it's gonna cause some merge conflicts

Contributes to #55729

- references
https://www.conventionalcommits.org/en/v1.0.0/
https://regex101.com/

Some of these have a lot of labels, check the edits for a timeline
- testing
https://github.com/casswedson/auto-label-issue-pr/issues/8
https://github.com/casswedson/auto-label-issue-pr/issues/3
https://github.com/casswedson/auto-label-issue-pr/pull/2

### CODE REVIEW

File: .github/labeler.yml [hunk 1]
```diff
@@ -1,51 +1,62 @@
...
+labels:
+  '<Bugfix>':
+    include:
+      - '\b^fix(\(.*\))?(!)?:(.*)?\b'
+
+  'Game: Balance':
+    include:
+      - '\b^balance(\(.*\))?(!)?:(.*)?\b'
+
+  '<Bug>':
```
Comment (anothersimulacrum) [inline .github/labeler.yml:24]: This should not be an autolabel, because it blocks stalebot.

File: .github/pr-labels.yml [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+  - "**/*.py"
+
+"Documentation":
+  - "**/*.md"
+  - doc/**/*
+
+"Code: Tooling":
+  - tools/**/*
+
+"Code: Test":
```
Comment (dseguin) [inline .github/pr-labels.yml:18]:
```suggestion
"Code: Tests":
```

Just noticed that the existing label wasn't being used:

labels

Comment (casswedson) [reply .github/pr-labels.yml:18]: Ah, yeah I was planning on fixing that, also include the changes from my other fix label box pr

File: .github/pull_request_template.md [hunk 3]
```diff
@@ -7,10 +7,49 @@ The results of these can be either seen under the "Files changed" section of a P
...
+- i18n or translation
+- maps
+- missions
+- mods
+- tools for code that is not part of the game but is part of the repo.
+
+Based on https://www.conventionalcommits.org/en/v1.0.0/
+-->
+
 #### Summary
```
Comment (kevingranade) [inline .github/pull_request_template.md:49]: The summary section needs to stay at the top, if there is one thing the submitter reads and acts on, this must be it, and everything else is optional.

**Comment by kevingranade** [comment]
I'm not sure about this one at all, it's a fairly complex syntax and I'm very skeptical of it triggering reliably, and if it's not reliable I don't want it because it will cause more work for triage rather than less.

**Comment by casswedson** [comment]
1. Regarding the docs for this
Yeah, I'll probably put that somewhere in the docs folder, front and center for something this complex was a bad idea
2. Complex syntax
It needs to be fairly complex, the more verbose and strict the regex the more reliable it would be
3. I'm very skeptical of it triggering reliably
I really can't tell you it's gonna work perfect from the start, in fact it's not gonna as demonstrated before I am stupid (also ignorant) and github automation can be pretty unpredictable, like it took around 5 prs to get the other auto labeler working fine, it ain't perfect, but it works 
I tested it here <https://github.com/casswedson/auto-label-issue-pr> however, a trail by fire will surely be needed to iron out the flaws this has, I am willing to stick around and fix it
4. If this just does not work the option to shut it down completely is open, quick revert, that's it

I have proofs of concept but as a demonstration I can do summary line based labeling first (I was planning to that anyway)

**Comment by kevingranade** [comment]
There's a dilemma here, the outcome of which is that I don't want to use this system.
The problem is for this to be useful, 100% of PRs have to have a full complement of entries, and that will not happen unless we make it mandatory and build even more tooling to fail PRs that don't comply.
The other side is do I think doing so is a reasonable ask for every contributor, and the answer is no.

I guarantee that as useful as this might be, it would prevent some contributors from completing the PR process because it's just a bit too much bureaucracy, and that's too high a cost for better labelling.

In a more general sense, this doesn't automate labeling, it just moves responsibility for doing it to the author, which is not something that IMO improves the situation.

**Pull request closed without merging**